hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/ProcessorTasklet.java

```diff
@@ -343,11 +344,29 @@ private void stateMachineStep() {
             case EMIT_DONE_ITEM:
                 if (outbox.offerToEdgesAndSnapshot(DONE_ITEM)) {
                     ssContext.processorTaskletDone();
-                    state = END;
-                    progTracker.done();
+                    progTracker.madeProgress();
+                    state = CLOSE;
+                    stateMachineStep();
                 }
                 return;
 
+            case CLOSE:
+                if (isCooperative()) {
+                    if (closeFuture == null) {
+                        closeFuture = executionService.submit(this::closeProcessor);
```

Comment:
I wonder if this will slow down job completion, since most processors will have a noop close implementation?

Reply:
Maybe we can have a no-op constant that we recognize and short-circuit this?

Reply:
@cangencer Using JMH submitting 100 no-op tasks to ForkJoinPool.commonPool and waiting for their completion by checking future.isDone takes on average 300μs. So it's not that much pain.

@mtopolnik It's not a lambda, but a method we call. We can check if the `declaringClass` of `close` method is the `Processor` class:
```java
if (p.getClass().getMethod("close").getDeclaringClass() == Processor.class)
    // ignore the close method
```
I benchmarked this too. This check took 1.3μs, while directly executing a no-op runnable took 1.7μs. I think it's not worth the complexity, we're not on a hot path.